keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+
+In practice above limitation should be fine. We will add events to PV and PVC when we detect that user has reduced requested size. Admin can recover from this state either by manually increasing quota or by manually rebuilding the PVC object.
+
+#### User flow stories
+
+###### Case 1 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+
+
+###### Case 2 (controller+node expandable):
+
+- User increases 10Gi PVC to 100Gi
+- Api Server sets `pvc.spec.allocatedResources` to 100Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
+- Expansion to 100Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 10Gi.
+- User requests size to 20Gi.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Spec.AllocatedResources` is `100Gi`.
+- Expansion succeeds and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `20Gi`.
+- `pvc.Spec.AllocatedResources` however keeps reporting `100Gi`.
+- User expands PVC to 120Gi
+- Api Server sets `pvc.spec.allocatedResources` to 120Gi.
+- Quota controller uses `max(pvc.Spec.AllocatedResources, pvc.Spec.Resources)` and adds `20Gi` to used quota.
+- Expansion to 120Gi fails and hence `pv.Spec.Capacity` and `pvc.Status.Capacity `stays at 20Gi.
+
+### Risks and Mitigations
+
+- One risk as mentioned above is, if expansion failed and user retried expansion(successfully) with smaller value, the quota code will keep reporting higher value. In practice though - this should be acceptable since such expansion failures should be rare and admin can unblock the user by increasing the quota or rebuilding PVC if needed. We will emit events on PV and PVC to alerts admins and users.
+
+## Graduation Criteria
+
+* Beta in 1.18 behind `ExpandPersistentVolumes`(currently in beta) feature gate. We are not planning to introduce a new feature gate
+for this feature. But if need be, we may introduce this as an alpha feature.
+* GA in 1.19 or 1.20 depending on this feature being beta or alpha in 1.18.
+
+### Test Plan
+
+* Basic unit tests for storage strategy of PVC and quota system.
+* E2e tests using mock driver to cause failure on expansion and recovery.
```

Comment:
should this be added to the external test suite so all CSI drivers can also be tested against this